packages/ckeditor5-engine/src/conversion/downcastdispatcher.js

```diff
@@ -489,6 +540,156 @@ export default class DowncastDispatcher {
+	 *
+	 * @param {module:engine/model/differ~Differ} differ The differ object with buffered changes.
+	 * @returns {Array.<Object>}
+	 * @private
+	 */
+	_getChangesAfterAutomaticRefreshing( differ ) {
+		const changes = differ.getChanges();
+		const elementsToRefresh = this._getElementsForAutomaticRefresh( changes );
+
+		if ( !elementsToRefresh.size ) {
+			return changes;
+		}
+
+		for ( const element of elementsToRefresh.values() ) {
+			differ.refreshItem( element );
+		}
+
+		// The `differ.refreshItem()` invalidates differ cache - we can't re-use previous changes.
+		const changesAfterRefresh = differ.getChanges();
+
+		return changesAfterRefresh.filter( entry => !elementsToRefresh.has( getElementFromChange( entry ) ) );
+	}
+
+	/**
+	 * Returns elements that should be converted using {@link #convertRefresh} defined by a `triggerBy` configuration for
+	 * {@link module:engine/conversion/downcasthelpers~DowncastHelpers#elementToElement `elementToElement()`} conversion helper.
+	 *
+	 * This will return:
+	 *
+	 * * Element which attributes changed for an 'attribute' change.
+	 * * Parent of inserted or removed element for matched 'insert' or 'remove' changes.
+	 *
+	 * @param {Array.<Object>} changes The changes diff set from the differ.
+	 * @returns {Set.<module:engine/model/element~Element>}
+	 * @private
+	 */
+	_getElementsForAutomaticRefresh( changes ) {
+		const found = changes
+			.filter( ( { type } ) => type === 'attribute' || type === 'insert' || type === 'remove' )
+			.map( entry => {
+				const element = getElementFromChange( entry );
+				let eventName;
+
+				if ( entry.type === 'attribute' ) {
+					if ( !element ) {
+						// Refreshing is done only on elements so skip text attribute changes.
+						return;
+					}
+
+					eventName = `attribute:${ entry.attributeKey }:${ element.name }`;
+				} else {
+					eventName = `${ entry.type }:${ entry.name }`;
+				}
+
+				if ( this._refreshTriggerEventToElementNameMapping.has( eventName ) ) {
+					return element;
+				}
+			} )
+			.filter( element => !!element );
+
+		return new Set( found );
+	}
+
+	/**
+	 * Handles reconverting a model element that has an existing model-to-view mapping.
+	 *
+	 * It performs a shallow conversion for the element and its attributes. All children that already have a converted view
+	 * will not be converted again. Their existing view elements will be used instead.
+	 *
+	 * @private
+	 * @param {Object} data Event data.
+	 */
+	_reconvertElement( data ) {
+		// Cache current view element of a converted element, might be undefined if first insert.
+		const currentView = this.conversionApi.mapper.toViewElement( data.item );
+
+		// Remove the old view (but hold the reference as it will be used to bring back view items not needed to re-render.
+		// Thanks to the mapper that holds references nothing should blow up.
+		this.conversionApi.writer.remove( currentView );
```

Comment:
`currentView` is used only here, later in code we use an element from mapper, so this code comment seems invalid.